File: ruby/prime-factors/prime_factors.rb
module PrimeFactors
  def self.for(number)
    [].tap do |factors|
      while number > 1
        factor = minimal_factor_of number
        number /= factor
        factors << factor
      end
    end
  end

  def self.minimal_factor_of(number)
    (2..).find { |factor| (number % factor).zero? }
  end
end

File: javascript/prime-factors/prime-factors.js
export const primeFactors = number => {
  const factors = [];

  while (number > 1) {
    for (const prime of primeNumbers()) {
      if (number % prime === 0) {
        number /= prime;
        factors.push(prime);
        break;
      }
    }
  }

  return factors;
};

function* primeNumbers() {
  const primes = [];
  let current = 2;

  while (true) {
    primes.push(current);
    yield current;

    let isPrime = false

    while (!isPrime) {
      current++;
      isPrime = true;

      for (const prime of primes) {
        if (current % prime === 0) {
          isPrime = false;
          break;
        }
      }
    }
  }
}
